Support not equal operator in song select search

Valerus9 opened the pull request:
Implements not equal operator and adds test coverage for it. Closes #34569


https://github.com/user-attachments/assets/b3dfde62-544d-45fc-af60-6d504e33f6f2

Valerus9 committed:
Implement not equal operator

Added blank lines

Valerus9 committed:
Add test coverage

bdach reviewed (changes requested):
only the parsing side is tested, missing test coverage for the actual evaluating side (`FilterMatchingTest`)

Valerus9 committed:
Rename ExludeIfInRange to InvertRange, add xmldox to InvertRange

Valerus9 committed:
Simplify logic of IsInRange()

Valerus9 commented on osu.Game/Screens/Select/FilterCriteria.cs:
```diff
@@ -158,21 +158,36 @@ public bool IsInRange(T value)
+                    if (minComparison == 0 && IsLowerInclusive && ExcludeIfInRange)
+                        return false;
+
+                    if (maxComparison == 0 && IsUpperInclusive && ExcludeIfInRange)
```
Double checked and implemented

Valerus9 committed:
Added FilterMatchTest for InvertRange

Valerus9 committed:
Fix code quality

Fix code quality 2

code quality fix 3

Fix code quality 4

Add empty line

bdach reviewed (changes requested):
From other notes:

- https://github.com/ppy/osu/blob/afe08a9476e7d8529ac232dd12dab98994a38e20/osu.Game/Screens/Select/FilterQueryParser.cs#L590-L597 needs updating to explicitly say that inequality isn't supported for the same reason that equality isn't.
- `played=0` and `played!=0` return the same results. You'll either want an `op == Operator.Equal` check here: https://github.com/ppy/osu/blob/afe08a9476e7d8529ac232dd12dab98994a38e20/osu.Game/Screens/Select/FilterQueryParser.cs#L75 or negate the `played` bool appropriately.

Valerus9 committed:
Add tests coverage for filters with not equal

Valerus9 commented on osu.Game/Screens/Select/FilterQueryParser.cs:
```diff
@@ -763,6 +829,32 @@ private static bool tryUpdateRankedDateRange(ref FilterCriteria.OptionalRange<Da
+                        minDateTimeOffset = dateTimeOffsetFromDateOnly(year.Value, month.Value, day.Value);
+                        maxDateTimeOffset = dateTimeOffsetFromDateOnly(year.Value, month.Value, day.Value).AddDays(1);
+                        return tryUpdateCriteriaRange(ref dateRange, Operator.Less, minDateTimeOffset)
+                               || tryUpdateCriteriaRange(ref dateRange, Operator.GreaterOrEqual, maxDateTimeOffset);
```
All of the filters now have tests (obviously if the filters are the same type then I wouldn't do redundant tests)

bdach committed:
Fix tests

- Move them to the correct class. They were exercising filter matching,
  not parsing.
- Remove a bunch of unreadable tuple stuff that was mostly obfuscating
  the readability without actually improving test coverage.
- Make tests fail everywhere rather than on CI only. They were failing
  because they were written in a way that was implicitly dependent on
  the local computer's timezone.

bdach committed:
Fix date range filter being inexplicably implemented a little different than the equality filter

Why???

bdach commented:
> Any idea why this occours?

The tests are failing because you inadvertently made them dependent on the local computer's timezone, which I can excuse, but also... because you *for whatever reason* made a slight change to how the inequality filter works compared to the equality filter (https://github.com/ppy/osu/pull/34568/commits/7a2b032b00e4a261341701aa15049a3d4aaa1a3d) which as far as I can tell just makes it completely incorrect? Why would you do that change?

I'm not done with this yet mind, just wanting to get the most pressing pains out of the way...

bdach committed:
Reduce code duplication

bdach committed:
Fix code quality

bdach committed:
Refactor string filter not-equal implementation to not make my eyes bleed

bdach committed:
Remove redundant test

bdach committed:
Improve xmldoc

bdach committed:
Simplify multi-valued not-equal enum filter implementation (and cover with tests)

bdach reviewed (dismissed):
probably fine except for the part where I basically rewrote this from scratch, so going to need one more review on this from someone else

also we're squashing this one because the history here is a travesty

Valerus9 commented:
> The tests are failing because you inadvertently made them dependent on the local computer's timezone, which I can excuse
For future reference how do I prevent this from happening?

> because you for whatever reason made a slight change to how the inequality filter works compared to the equality filter (https://github.com/ppy/osu/commit/7a2b032b00e4a261341701aa15049a3d4aaa1a3d) which as far as I can tell just makes it completely incorrect? Why would you do that change?

Apperently my "local computer's timezone" decided that for whatever reason the next day (22 instead of 21) not equal with the time I give it in the test. And since I didn't know that I somehow used my "local computer's timezone" I thought that I had to modify the not equal's day check to fix it.

bdach committed:
Fix incorrect parsing

frenzibyte committed:
Merge branch 'master' into notequaloperator
